app/feature/feature-edit-coinsured/src/main/kotlin/com/hedvig/android/feature/editcoinsured/data/FetchCoInsuredPersonalInformationUseCase.kt

```diff
@@ -0,0 +1,39 @@
+package com.hedvig.android.feature.editcoinsured.data
+
+import arrow.core.Either
+import arrow.core.raise.either
+import arrow.core.raise.ensureNotNull
+import com.apollographql.apollo3.ApolloClient
+import com.hedvig.android.apollo.safeExecute
+import com.hedvig.android.apollo.toEither
+import com.hedvig.android.core.common.ErrorMessage
+import octopus.PersonalInformationQuery
+
+interface FetchCoInsuredPersonalInformationUseCase {
+  suspend fun invoke(ssn: String): Either<ErrorMessage, CoInsured>
+}
+
+class FetchCoInsuredPersonalInformationUseCaseImpl(
+  private val apolloClient: ApolloClient,
+) : FetchCoInsuredPersonalInformationUseCase {
+  override suspend fun invoke(ssn: String): Either<ErrorMessage, CoInsured> = either {
+    val result = apolloClient.query(PersonalInformationQuery(ssn))
+      .safeExecute()
+      .toEither(::ErrorMessage)
+      .bind()
+
+    ensureNotNull(result.personalInformation) {
+      ErrorMessage("No personal information found")
+    }
+
+    CoInsured(
+      firstName = result.personalInformation.firstName,
+      lastName = result.personalInformation.lastName,
+      birthDate = null,
```

Comment:
My thinking here was that these use cases will only be used in this module, and that we always work with this "CoInsured" data type in this module. Is this coupling too tight do you think? Should we return another datatype from this use case and it to CoInsured in the presenter instead?

Reply:
Yeah it's good that it's contained in the module, but what I just saw here is that in reality this doesn't want to return CoInsured, this just wants to return first name + last name, because that's what it fetches. Then the fact that birthDate or hasMissingInfo are what they are, is just a consequence of returning a CoInsured, but in reality it doesn't know anything about those. The call site might know that this is needed only for cases where we don't know those other two anyway, but it is a coincidence.

To put it in another scenario, if you have a CoInsured already, where you have their ID and you also know that their birthDate is 2000-01-01.
Then you'll have something like this to properly merge the two CoInsured objects
```
val existingCoInsured = CoInsured(
  id = "id",
  firstName = null,
  lastName = null,
  birthDate = "2000-01-01",
  hasMissingInfo = true,
)
val fetchedCoInsured: CoInsured = fetchCoInsuredPersonalInformationUseCase().invoke(existingCoInsured.id).getOrNull()!!
val resultingCoInsured = existingCoInsured.copy(
  id = "id",
  firstName = fetchedCoInsured.firstName,
  lastName = fetchedCoInsured.lastName,
  birthDate = "2000-01-01",
  hasMissingInfo = true,
)
```
This puts the burden of knowing that the use case returns a CoInsured, but actually only the first name and last name are the two important things, and you *should not* use the other fields, as they are just some sane defaults, they don't mean more than that.

The UseCase can return `data class CoInsuredPersonalInformation(val firstName: String, val lastName: String)` instead, and then you got an easy way of know that well you just got the first name + last name, do whatever you wish with it, there's no way to use the wrong information coming from it.

Reply:
Makes sense, changing 👍